src/lock/optiql.cpp

```diff
@@ -194,11 +231,33 @@ OptiQL::LockX()  //
 
 void
 OptiQL::UnlockX(  //
-    const uint64_t qid)
+    const uint64_t qid,
+    const uint64_t ver)
 {
   auto *qnode = &(_qnodes[qid]);
 
-  throw std::runtime_error{"not implemented yet"};
+  auto *next_ptr = qnode->next.load(kAcquire);
+  if (next_ptr == nullptr) {  // this is the tail node
+    auto cur = lock_.load(kRelaxed);
+    while (((cur & kQIDMask) >> kQIDShift) == qid) {
+      if (lock_.compare_exchange_weak(cur, ver, kRelease, kRelaxed)) {
+        RetainQID(qid);
+        return;
+      }
+      CPP_UTILITY_SPINLOCK_HINT
+    }
+  }
+
+  // enable opportunistic read
+  lock_.fetch_or(kOPReadLock | ver, kRelaxed);
```

Comment:
他のスレッドに変更内容を読ませる必要があるので，ここはフェンスがいります．

```suggestion
  lock_.fetch_or(kOPReadLock | ver, kRelease);
```